### DISCUSSION TITLE
KeyObject to JWK with v3

### DISCUSSION THREAD

**Discussion by bespokebob** [panva/jose#109] [Ideas]
**Is your feature proposal related to a problem? Please describe.**

In v3, there does not seem to be any way to generate and export a key as a JWK. There is a utility to parse a JWK into a native key, but not the other way around.

**Describe the solution you'd like**

An `exportJwk` that takes `KeyObject | CryptoKey` and returns `JWK`

**Describe alternatives you've considered**
 
Stay on v2, or copy the `keyObjectToJWK` and supporting utils

**Additional context**

The Web Crypto API natively supports parsing and exporting keys as JWK. The Node Crypto API does not, though.

- [x] i have searched the configuration section for this feature and couldn't find it
- [x] i have searched the issues tracker on github for similar requests and couldn't find anything related.

**Comment by panva** [comment]
See this is tricky, since Web Cryptography API has this affordance i don't feel like putting this in for the other runtime.

How would you feel if, instead, there was a separate, node-specific module for KeyObject to JWK conversion?

**Comment by bespokebob** [comment]
> How would you feel if, instead, there was a separate, node-specific module for KeyObject to JWK conversion?

I could live with that. I understand you're trying to keep the v3 API slimmed down (and coming from `node-jose`, I appreciate it). Though for users only importing the specific sub-modules they need, it doesn't seem like there would be a disadvantage to just including it in `jose`. It shouldn't increase the bundle size for anyone who doesn't need it in the browser.

**Comment by panva** [comment]
That's a fair point. I'll see what can be done. In the meantime, there are established modules for `pem -> jwk` conversion and since you have a keyObject at hand (see [keyobject.export](https://nodejs.org/api/crypto.html#crypto_keyobject_export_options)), you can use those for the time being.

**Comment by bespokebob** [comment]
> That's a fair point. I'll see what can be done. In the meantime, there are established modules for `pem -> jwk` conversion and since you have a keyObject at hand (see [keyobject.export](https://nodejs.org/api/crypto.html#crypto_keyobject_export_options)), you can use those for the time being.

That's a good suggestion. Exporting the key and using https://npm.io/package/rsa-pem-to-jwk does what I need.

**Comment by panva** [comment]
@youngbob i think https://www.npmjs.com/package/@trust/keyto is a more general alternative, supporting more key types.

**Comment by panva** [comment]
Nevertheless, i'll put a util function in for KeyLike to JWK conversion

**Comment by fidgi** [comment]
Hi,
May be this is not related but I encounter the following issue with version 3.0.2 : 
```
node_modules/jose/src/jwks/remote.ts:1:15 - error TS2305: Module '"crypto"' has no exported member 'KeyObject'.

1 import type { KeyObject } from 'crypto'
                ~~~~~~~~~

```
FYI : I'm trying to use the function (createRemoteJWKSet)[https://github.com/panva/jose/blob/master/docs/functions/_jwks_remote_.createremotejwkset.md]

Node Version used on a debian Linux  : 
$ node --version
v12.19.0

**Comment by bespokebob** [comment]
> Hi,
> May be this is not related but I encounter the following issue with version 3.0.2 : 
> ```
> node_modules/jose/src/jwks/remote.ts:1:15 - error TS2305: Module '"crypto"' has no exported member 'KeyObject'.
> 
> 1 import type { KeyObject } from 'crypto'
>                 ~~~~~~~~~
> 
> ```
> FYI : I'm trying to use the function (createRemoteJWKSet)[https://github.com/panva/jose/blob/master/docs/functions/_jwks_remote_.createremotejwkset.md]
> 
> Node Version used on a debian Linux  : 
> $ node --version
> v12.19.0

You probably need to install `@types/node`

**Comment by panva** [comment]
https://github.com/panva/jose/releases/tag/v3.1.0

**Comment by bespokebob** [comment]
> https://github.com/panva/jose/releases/tag/v3.1.0

Perfect, thanks!

**Comment by panva** [comment]
Please consider [supporting](https://github.com/panva/jose?sponsor=1) the library if it provides value to you or your company and this support was of help to you. Supporting the library means, amongst other things, that such support will be available in the future.
